casper/src/main/scala/coop/rchain/casper/util/comm/ApproveBlockProtocol.scala

```diff
@@ -0,0 +1,131 @@
+
+    validSig.fold(Log[F].warn(s"APPROVAL: ignoring invalid block approval from $sender")) { sig =>
+      sigsF.update(_ + sig) *> Log[F].info(s"APPROVAL: received block approval from $sender")
+    }
+  }
+
+  def currentSigs: F[Set[Signature]] = sigsF.get
+
+  def approvedBlock: F[Option[ApprovedBlock]] =
+    sigsF.get.map(
+      sigs =>
+        if (sigs.size >= requiredSigs) Some(ApprovedBlock(Some(candidate), sigs.toSeq))
+        else none[ApprovedBlock])
+
+  //TODO: potential optimization, only send to peers we have not
+  //      reveived a valid signature from yet
+  def sendUnapprovedBlock: F[Unit] =
+    for {
+      _ <- Log[F].info(s"APPROVAL: Beginning send of UnapprovedBlock $candidateHash to peers...")
+      _ <- CommUtil.sendToPeers[F](transport.UnapprovedBlock, serializedUnapprovedBlock)
+      _ <- Log[F].info(s"APPROVAL: Sent UnapprovedBlock $candidateHash to peers.")
+    } yield ()
+
+  def sendApprovedBlock: F[Boolean] =
+    approvedBlock.flatMap {
+      case Some(a) =>
+        val serializedApprovedBlock = a.toByteString
+        for {
+          _ <- Log[F].info(s"APPROVAL: Beginning send of ApprovedBlock $candidateHash to peers...")
+          _ <- CommUtil.sendToPeers[F](transport.ApprovedBlock, serializedApprovedBlock)
+          _ <- Log[F].info(s"APPROVAL: Sent ApprovedBlock $candidateHash to peers.")
+        } yield true
+
+      case None => false.pure[F]
+    }
+
+}
+
+object ApproveBlockProtocol {
+
+  def apply[F[_]](implicit instance: ApproveBlockProtocol[F]): ApproveBlockProtocol[F] = instance
+
+  def create[F[_]: Capture: Sync: NodeDiscovery: TransportLayer: Log: Time: ErrorHandler](
+      block: BlockMessage,
+      requiredSigs: Int,
+      duration: Long): F[ApproveBlockProtocol[F]] =
+    for {
+      now   <- Time[F].currentMillis
+      sigsF <- Ref.of[F, Set[Signature]](Set.empty[Signature])
+    } yield new ApproveBlockProtocol(block, requiredSigs, now, duration, sigsF)
+
+  def blockApprovalPacketHandler[F[_]: Monad: ApproveBlockProtocol](
+      peer: PeerNode): PartialFunction[Packet, F[Option[Packet]]] =
+    Function
+      .unlift(packetToBlockApproval)
+      .andThen {
+        case ba: BlockApproval =>
+          for {
+            _ <- ApproveBlockProtocol[F].addApproval(ba)
+          } yield none[Packet]
+      }
+
+  def run[F[_]: Monad: Capture: Time: ApproveBlockProtocol](waitTime: Long = 5000L): F[Unit] =
+    for {
+      t             <- Time[F].currentMillis
+      maybeApproved <- ApproveBlockProtocol[F].approvedBlock
+      _ <- maybeApproved match {
+            case Some(_)
+                if (ApproveBlockProtocol[F].candidate.requiredSigs == 0 || t >= ApproveBlockProtocol[
+                  F].start + ApproveBlockProtocol[F].duration) =>
+              ApproveBlockProtocol[F].sendApprovedBlock
+
+            case _ =>
+              ApproveBlockProtocol[F].sendUnapprovedBlock *> IOUtil.sleep(waitTime) *> run[F](
```

Comment:
Hmm, I think `IOUtil.sleep(…)` is not a proper way to do sleep. IMO we should use [`cats-effect.Timer#sleep`](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Timer.scala#L154) and [the implementation](https://github.com/typelevel/cats-effect/blob/master/core/jvm/src/main/scala/cats/effect/internals/IOTimer.scala#L47).

Reply:
Make a PR to change it? I used what was already in `shared` but don't really have strong opinions on this.

Reply:
I will, I am just making notes for both you and myself for the future.